BUG: Use stable algorithm for _nanvar.

jvkersch opened the pull request:
closes #10242

This PR replaces the sum-of-squares algorithm used to compute the variance by a more stable algorithm. The algorithm here is essentially the same as used in numpy 1.8 and up, and I've added a TODO to replace the implementation with a direct call to numpy when that version is the default.

Somewhat counter to the discussion in #10242, I chose not to go with the Welford algorithm, for two reasons: numpy, and the fact that `_nanvar` needs to be able to deal with arrays of different shape, which is tricky to get right in Cython.

Commit:
BUG: Use stable algorithm for _nanvar.

jvkersch commented:
@shoyer or @jreback Would either of you mind giving this a quick review? Thanks!

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -339,6 +339,13 @@ def test_nanmean_overflow(self):
+    def test_nanstd_roundoff(self):
+        # Regression test for GH 10242 (test data taken from GH 10489). Ensure
+        # that variance is stable.
+        data = Series(766897346 * np.ones(10))
```
can u add a couple of more tests with some fixed values and a fixed result (that is not 0)

also included an example with nans

TomAugspurger commented:
So the plan is to replace this with numpy's version once we bump to that to 1.8? Could you open up a separate issue as to reminded us to do that?

jreback commented:
we won't be eliminating 1.7 for quite some time FYI 
so this would have to be a conditional (and could even be out in now with the conditional on the version)

shoyer commented:
This looks pretty reasonable to me.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -397,4 +397,4 @@ Bug Fixes
 - Bug in vectorised setting of timestamp columns with python ``datetime.date`` and numpy ``datetime64`` (:issue:`10408`, :issue:`10412`)
 - Bug in ``pd.DataFrame`` when constructing an empty DataFrame with a string dtype (:issue:`9428`)
-
+- Bug in ``_nanvar`` causing inaccurate results due to unstable algorithm  (:issue:`10242`)
```
Bug in``.var()`` no-one knows (or cares how this is implemented). Be much more explict on when this occurs. E.g. highly similar values. This sounds too alarmist.

jvkersch replied:
Done.

jreback commented on pandas/core/nanops.py:
```diff
@@ -353,10 +353,16 @@ def _nanvar(values, axis=None, skipna=True, ddof=1):
-    result = np.fabs((XX - X * X / count) / d)
-    return result
+    # This implementation is based on that of nanvar in numpy 1.8 and up. TODO:
+    # call directly into numpy when we bump numpy up 1.8 .
```
put the issue number as a comment here. I would take out the reference, unless you want to add a conditional if the user has >= 1.8 and use the function, otherwise I suspect this wont' be added.

jvkersch replied:
I added the conditional to `nanvar` itself, but I just realized that I didn't check whether anything calls `_nanvar` directly. In any case, I don't have any strong preferences for putting the conditional in any particular place.

jreback commented on pandas/core/nanops.py:
```diff
@@ -353,10 +353,16 @@ def _nanvar(values, axis=None, skipna=True, ddof=1):
+        avg = np.expand_dims(avg, axis)
+    sqr = _ensure_numeric((avg - values) ** 2)
+    np.putmask(sqr, mask, 0)
+    var = sqr.sum(axis=axis) / d
```
pass ``dtype='float64'``. I think you always need to upcast here (not sure though, maybe can add some tests for this).

jvkersch replied:
What exactly is the purpose of the upcast? If it is to avoid overflowing when we take the sum and the square, it should probably go at the beginning of the computation.

jreback replied:
we always specify the ``dtype=`` kw when summing (even for float), because this is the accumulator used. try this with large ``float32`` values and see what happens. (yes this should be on both ``.sum(..)``

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -339,6 +339,13 @@ def test_nanmean_overflow(self):
+        data = Series(766897346 * np.ones(10))
+        result = data.std()
+        self.assertEqual(result, 0.0)
+
```
need some tests for ddf 0/1 as well

jreback commented on pandas/core/nanops.py:
```diff
@@ -353,10 +353,16 @@ def _nanvar(values, axis=None, skipna=True, ddof=1):
-    XX = _ensure_numeric((values ** 2).sum(axis))
-    result = np.fabs((XX - X * X / count) / d)
-    return result
+    # This implementation is based on that of nanvar in numpy 1.8 and up. TODO:
```
pls add an exaplanation of the algo as well (and that its stable)

Commit:
BUG: Allow exceptions to propagate to top level.

A missing 'raise' caused any exception in check_fun_ddof to be
ignored. Restoring the raise revealed some issues with the existing
tests:

1. Optional keyword arguments were not passed into the function to be
   tested.
2. nanvar, nanstd, nansem were being tested on string data, but are
   not able to handle string inputs.
3. test_nansem compared the output of nanops.nansem to numpy.var, which
   probably should have been scipy.stats.sem, judging from the
   conditional at the top of the test.

This commit also replaces all BaseExceptions by checks for AssertionError.

jvkersch commented:
Thanks all for the extensive comments!

@jreback I made a first pass at addressing your comments. Regarding the relative lack of tests, I assumed that the existing tests in `test_nanops` would be sufficient to check the behavior of the function in the face of NaNs, different ddofs, etc. However, I noticed that the test infrastructure in `test_nanops` was missing a `raise`, causing some test errors to be swallowed.  I can make that commit into a separate PR if necessary.

jreback commented on pandas/core/nanops.py:
```diff
@@ -373,7 +383,16 @@ def nanvar(values, axis=None, skipna=True, ddof=1):
+    else:
+        np_nanvar = getattr(np, 'nanvar', None)  # Present in Numpy >= 1.8
+        if np_nanvar is None:
+            return _nanvar(values, axis=axis, skipna=skipna, ddof=ddof)
```
I would rather put ``_nanvar`` here. Unless we are calling it for another reason?

jreback commented:
@jvkersch I would like to see an explicty test that checks against a hard value (with an w/o nans), just to lock this down (rather than computing agains ``np.var`` which is what we do IIRC). That is generally fine, but since you are addressing this, can't hurt.

jvkersch commented:
Just a quick note: I haven't forgotten about this, but I'm currently lacking the time to finish this off. What's hindering me a bit is the lack of uniformity in the api for the nanops: some cast their arguments to float64 right away, some only do that for int dtypes, ... there's the implicit assumption that the nanop will raise `TypeError` for non-numeric input, etc. I think it might be worth fleshing out this api in a future PR. If I can't get things to work in a satisfactorily manner I'll just go for the minimally viable implementation here, and work on that instead, before revisiting.

jvkersch commented:
@jreback Would you be able to give this another review? Since last time, I took care of your earlier comments (explicit tests, folding `_nanvar` into `nanvar`, using a wide accumulator), but some unexpected side-effects plagued me a little:

1. Since `nanstd` called `_nanvar` directly, having it call `nanvar` instead meant fixing up some tests.
2. Numpy generates a RuntimeWarning whenever the number of dofs is not strictly positive. To avoid being swamped by those warnings, I turned off RuntimeWarnings when calling Numpy's `var` or `nanvar`.
3. Some tests for `numeric_only` called `nanvar` with non-numeric data and listened for a `TypeError`. This happened to work with the earlier implementation, but in some cases the new implementation raises a `ValueError` instead. Implementation-wise this works fine, too, but one test needed to be made a little more flexible.

Thanks!

jreback commented on pandas/core/nanops.py:
```diff
@@ -338,12 +339,38 @@ def _get_counts_nanvar(mask, axis, ddof, dtype=float):
+
+
+@disallow('M8')
+@bottleneck_switch(ddof=1)
```
why are you changing this. No real need to go to numpy at all. Either its hit in bottleneck, or we have our own algo which is modelled on ``nanvar``.

jvkersch replied:
@jreback I think I misunderstood one of your earlier suggestions. To clarify, are you suggesting that we don't go to Numpy at all in `nanvar`, and just have our own implementation (which used to be `_nanvar`)? I do think that would simplify matters a bit.

jreback replied:
yep

jvkersch replied:
@jreback Done (and squashed)

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -817,6 +831,99 @@ def test_non_convertable_values(self):
+    def test_nanvar_all_finite(self):
+        samples = self.samples
+        actual_variance = nanops.nanvar(samples)
+        np.testing.assert_almost_equal(
```
use ``tm.assert_almost_equal`` or ``tm.assert_numpy_array_equal`` as appropriate

jvkersch replied:
I resorted to using the Numpy variants because I needed to set the decimal precision to something low; is there a Pandas version that also allows the precision to be adjusted?

jreback replied:
I think we have a less_precise option where u can set the decimal places to compare (or defaults to 3 I think)

jvkersch replied:
IIUC the number of decimal places is fixed to 5 or 3 (depending on whether `check_less_precise` is True) and I really need 2 ;-)  I'll see if I can work around this.

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -817,6 +831,99 @@ def test_non_convertable_values(self):
+        actual_variance = nanops.nanvar(samples, skipna=False)
+        np.testing.assert_almost_equal(
+            actual_variance, np.nan, decimal=2)
+
```
test with ``nanstd`` as well for all-nan

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -817,6 +831,99 @@ def test_non_convertable_values(self):
+        samples = np.nan * np.ones(2 * self.samples.shape[0])
+        samples[::2] = self.samples
+
+        actual_variance = nanops.nanvar(samples, skipna=True)
```
this should fail on numpy 1.7 because ``nanvar`` doesn't exist., and yet it doesn't. The travis 2.7 builds are all with 1.7

jvkersch replied:
This is using Pandas' `nanvar` ;-)

jreback replied:
oh right ok

jreback commented on pandas/tests/test_nanops.py:
```diff
@@ -710,7 +724,7 @@ def test__has_infs(self):
                 self.check_bool(nanops._has_infs, val.astype('f4'), correct)
                 self.check_bool(nanops._has_infs, val.astype('f2'), correct)
-            except BaseException as exc:
+            except AssertionError as exc:
```
why are you changing all of these?

jvkersch replied:
You're right, that was braindead. Reverted.

jvkersch commented:
@jreback I've added some tests for `nanstd` (though they are not as extensive as the `nanvar` tests). I couldn't get things to work reliably with `tm.assert_array_equal` (because of the fixed precision) so I'm hoping this can stay as-is (and I wouldn't mind revisiting this after #10788 has been addressed). Does this look good to you?

jreback commented:
looks good. pls rebase/squash. ping when green.

jvkersch commented:
@jreback Done!

jreback commented:
merged via https://github.com/pydata/pandas/commit/2d0d58c04114b634072e9425824b34cda3aeaef5

@jvkersch awesome job!

thanks!

jvkersch commented:
@jreback Thanks for the swift and thorough review!
